pincer/commands/chat_command_handler.py

```diff
@@ -0,0 +1,435 @@
+
+                # `key` represents the hash value for the top-level command that will
+                # hold the subcommand.
+
+                key = _hash_app_command_params(
+                    cmd.group.name,
+                    cmd.app.guild_id,
+                    AppCommandOptionType.SUB_COMMAND,
+                    None,
+                    None
+                )
+
+                if key not in ChatCommandHandler.built_register:
+                    ChatCommandHandler.built_register[key] = AppCommand(
+                        name=cmd.group.name,
+                        description=cmd.group.description,
+                        type=AppCommandOptionType.SUB_COMMAND,
+                        guild_id=cmd.app.guild_id,
+                        options=[]
+                    )
+
+                # No checking has to be done before appending `cmd` since it is the
+                # lowest level.
+                ChatCommandHandler.built_register[key].options.append(
+                    AppCommandOption(
+                        name=cmd.app.name,
+                        description=cmd.app.description,
+                        type=AppCommandType.CHAT_INPUT,
+                        options=cmd.app.options
+                    )
+                )
+
+                continue
+
+            # All single-level commands are registered here.
+            ChatCommandHandler.built_register[
+                _hash_app_command(cmd.app, cmd.group, cmd.sub_group)
+            ] = cmd.app
+
+    @staticmethod
+    def get_local_registered_commands() -> ValuesView[AppCommand]:
+        return ChatCommandHandler.built_register.values()
+
+    async def __get_existing_commands(self):
+        """|coro|
+
+        Get AppCommand objects for all commands registered to discord.
+        """
+        try:
+            self._api_commands = await self.get_commands()
+
+        except ForbiddenError:
+            logging.error("Cannot retrieve slash commands, skipping...")
+            return
+
+    async def __remove_unused_commands(self):
+        """|coro|
+
+        Remove commands that are registered by discord but not in use
+        by the current client
+        """
+        local_registered_commands = self.get_local_registered_commands()
+
+        def should_be_removed(target: AppCommand) -> bool:
+            for reg_cmd in local_registered_commands:
+                # Commands have endpoints based on their `name` amd `guild_id`. Other
+                # parameters can be updated instead of deleting and re-registering the
+                # command.
+                if (
+                    target.name == reg_cmd.name
+                    and target.guild_id == reg_cmd.guild_id
+                ):
+                    return False
+            return True
+
+        # NOTE: Cannot be generator since it can't be consumed due to lines 743-745
+        to_remove = [*filter(should_be_removed, self._api_commands)]
```

Comment:
`should_be_removed` is used once within a filter, then converted back to a list, could make that a list comprehension, right?

Reply:
```suggestion
to_remove = [
    cmd for cmd in self._api_commands
    if all(
        (
            target.name != reg_cmd.name
            or target.guild_id != reg_cmd.guild_id
        ) for reg_cmd in local_registered_commands
    )
]
```

Reply:
not a problem for this PR